latch/registry/record.py

```diff
@@ -1,20 +1,97 @@
     id: str
     name: str
 
-    _value_dict: Dict[str, Any]
+    _values: Dict[str, object]
+    _valid: bool = True
+
+    def get(self, key: str, default_if_missing: Optional[object] = None):
+        return self._values.get(key, default_if_missing)
+
+    def __getitem__(self, key: str):
+        if key not in self._values:
+            raise KeyError(f"column {key} not found in record {self.id} ({self.name})")
+        return self._values[key]
+
+    @classmethod
+    def from_id(cls, id: str):
+        data = execute(
+            """
+            query recordQuery($argRecordId: BigInt!) {
+                catalogSample(id: $argRecordId) {
+                    id
+                    name
+                    experiment {
+                        id
+                        removed
+                        catalogExperimentColumnDefinitionsByExperimentId {
+                            nodes {
+                                key
+                                type
+                            }
+                        }
+                    }
+                    catalogSampleColumnDataBySampleId {
+                        nodes {
+                            data
+                            key
+                        }
+                    }
+                }
+            }
+            """,
+            {"argRecordId": id},
+        )["catalogSample"]
+
+        if data.get("experiment") is None or data["experiment"]["removed"]:
+            return InvalidValue(json.dumps({"sampleId": id}))
+
+        record_data_dict = {
+            node["key"]: node["data"]
+            for node in data["catalogSampleColumnDataBySampleId"]["nodes"]
+        }
+
+        column_types_dict = {
+            node["key"]: node["type"]
+            for node in data["experiment"][
+                "catalogExperimentColumnDefinitionsByExperimentId"
+            ]["nodes"]
+        }
+
+        python_values = {}
+        valid = True
+
+        for key, registry_type in column_types_dict.items():
+            python_literal = registry_empty_cell
+            registry_literal = record_data_dict.get(key)
+            if registry_literal is not None:
+                python_literal = to_python_literal(
+                    registry_literal,
+                    registry_type["type"],
+                )
+            if registry_literal is None and not registry_type["allowEmpty"]:
+                valid = False
 
-    def get_value(self, key: str):
-        if key not in self._value_dict:
-            raise KeyError(f"Column {key} not found in record")
-        return self._value_dict[key]
+            python_values[key] = python_literal
 
-    def get_type(self, key):
-        if key not in self._type_dict:
-            raise KeyError(f"Column {key} not found in record")
-        return self._type_dict[key]
+        return cls(
+            id=id,
+            name=data["name"],
+            _values=python_values,
+            _valid=valid,
+        )
```

Comment:
I don't really like `valid` here since valid in our database only applies to primitive types. In this case, valid is only ever false if the column is `not allow empty` and the value is empty. I would probably get rid of `valid` here entirely.

Reply:
i added this to be similar to console where rows are marked invalid if any of their entries are invalid

Reply:
fixed bug appearing in two places where record entries are given `RegistryInvalidValue` values but the record wasn't marked invalid

Reply:
idk what the best thing is rn